plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -845,11 +845,49 @@ func PodFitsHostPorts(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *s
 	}
 
 	existingPorts := nodeInfo.UsedPorts()
-	for wport := range wantPorts {
-		if wport != 0 && existingPorts[wport] {
-			return false, []algorithm.PredicateFailureReason{ErrPodNotFitsHostPorts}, nil
+
+	// try to see whether two hostPorts will conflict or not
+	for existingPort, isUsed := range existingPorts {
+		if isUsed {
+			existingHostPortInfo := decode(existingPort)
+
+			if existingHostPortInfo.hostIP == schedutil.DefaultBindAllHostIP {
+				// loop through all the want hostPort to see if there exists a conflict
+				for wantPort := range wantPorts {
+					wantHostPortInfo := decode(wantPort)
+
+					// if there already exists one hostPort whose hostIP is 0.0.0.0, then the other want hostport (which has the same protocol and port) will not fit
+					if wantHostPortInfo.hostPort == existingHostPortInfo.hostPort && wantHostPortInfo.protocol == existingHostPortInfo.protocol {
+						return false, []algorithm.PredicateFailureReason{ErrPodNotFitsHostPorts}, nil
+					}
+				}
+			}
+		}
+	}
+
+	for wantPort := range wantPorts {
```

Comment:
Why we need 2 loops to check conflict? can you help to give a specific case?
My concern is the performance impact.

Reply:
I remembered Wizard once explained somewhere, seems need to be explained again :)

Reply:
wow, sorry, can you share the link of comment? I'd like to work together to improve that :).

Reply:
@k82cn it is all about checking special case for 0.0.0.0 hostIP. there are 2 cases for 2 loops:
1 if a new pod want to bind a hostPort on 0.0.0.0(like 0.0.0.0/80/TCP), and we need to find out if the node has exisiting port and protocol conflict to it.(0.0.0.0/80/TCP or 127.0.0.1/80/TCP both conflict)
2 if a new pod want to bind a hostPort on a node(127.0.0.1/80/TCP) and we need to find out if the node has existing port and protocol hostPort(0.0.0.0/80/TCP conflict).

Reply:
I understand the performance concern, I'm not quite sure about this is the best method, hoping to improve it with you too

Reply:
can we just store `0.0.0.0/80/TCP` in cache, and translate wantPot to `0.0.0.0/80/TCP` format?

Reply:
I don't quite understand...

Reply:
can you be more specific?

Reply:
only store `0.0.0.0/80/TCP` in cache; change `127.0.0.1/80/TCP` to `0.0.0.0/80/TCP` for wantPorts, and check against existingPorts

Reply:
1 how to decide when to  change 127.0.0.1/80/TCP to 0.0.0.0/80/TCP for wantPorts? I think it needs to only to change when existingPorts is 0.0.0.0/80/TCP.
2 what about your algorithm solving this case "there already exists one pod bind on 127.0.0.1/80/TCP and another pod wants to bind 0.0.0.0/80/TCP"?

Reply:
@k82cn PTAL

Reply:
re 1.), always change 127.0.0.1 to 0.0.0.0 for wantPorts; I think there'll be less wantPort in a Pod.

re 2.), we always record 0.0.0.0 into cache, and change 127.0.0.1 to 0.0.0.0 for wantPorts; so we just check it directly :).

does this make sense? The complexity is still O(wantPods * existPods), but we enhance the factor.

Reply:
@k82cn 1 we can't always change 127.0.0.1 to 0.0.0.0 for wantPorts, For example a want pod which uses (127.0.0.1/80/TCP) and a existing pod which uses (127.0.0.2/80/TCP), they can resides on same host, but according to your algorithm, they can't resides together.

 I come up with a method which uses another two map[string]struct{} (use space to trade for time)
```
wantmap map[string]struct{}
existingmap map[string]struct{} 
```
key is "port/protocol". Using this method can reduce inner for loop.
So the complexity is O(wantPods) + O(existingpods)

Reply:
>  a want pod which uses (127.0.0.1/80/TCP) and a existing pod which uses (127.0.0.2/80/TCP), they can resides on same host, but according to your algorithm, they can't resides together.

Why?

Reply:
I think it again and maybe above method is a little ugly, I come up with a new data structure
map[string]string, the key is "protocol/hostip/hostport", value is "protocol/hostport".
and delete the existing pod in the map instead of the old way(making it false).

Reply:
u change the 127.0.0.1/80/TCP to 0.0.0.0/80/TCP according to "always change 127.0.0.1 to 0.0.0.0 for wantPorts"

Reply:
I think they can resides in the same host according the algorithm.

Reply:
u mean the old algorithm which is using  map to check whether "protocol/hostIP/port" is in the map? 
how about your new algorithm "always change 127.0.0.1 to 0.0.0.0 for wantPorts; I think there'll be less wantPort in a Pod" for this case:
want port 127.0.0.1/80/tcp
existing port 127.0.0.1/80/tcp
They will be conflict if resides on the same host? can your algorithm detect that?